protocols/upnp/src/behaviour.rs

```diff
@@ -0,0 +1,515 @@
+    let duration = if permanent { 0 } else { MAPPING_DURATION };
+
+    P::add_port(
+        gateway,
+        mapping.protocol,
+        mapping.internal_addr,
+        Duration::from_secs(duration),
+    )
+    .map(move |result| match result {
+        Ok(()) => Event::Mapped(mapping),
+        Err(err) => Event::MapFailure(mapping, err),
+    })
+    .boxed()
+}
+
+/// Remove a port mapping on the gateway.
+fn remove_port_mapping<P: Provider + 'static>(
+    gateway: Arc<P::Gateway>,
+    mapping: Mapping,
+) -> BoxFuture<'static, Event> {
+    P::remove_port(gateway, mapping.protocol, mapping.internal_addr.port())
+        .map(move |result| match result {
+            Ok(()) => Event::Removed(mapping),
+            Err(err) => Event::RemovalFailure(mapping, err),
+        })
+        .boxed()
+}
+
+/// A [`Provider::Gateway`] event.
+#[derive(Debug)]
+enum Event {
+    /// Port was successfully mapped.
+    Mapped(Mapping),
+    /// There was a failure mapping port.
+    MapFailure(Mapping, Box<dyn Error>),
+    /// Port was successfully removed.
+    Removed(Mapping),
+    /// There was a failure removing the mapping port.
+    RemovalFailure(Mapping, Box<dyn Error>),
+}
+
+/// Mapping of a Protocol and Port on the gateway.
+#[derive(Debug, Clone)]
+struct Mapping {
+    listener_id: ListenerId,
+    protocol: Protocol,
+    multiaddr: Multiaddr,
+    internal_addr: SocketAddrV4,
+}
+
+impl Hash for Mapping {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.listener_id.hash(state);
+    }
+}
+
+impl PartialEq for Mapping {
+    fn eq(&self, other: &Self) -> bool {
+        self.listener_id == other.listener_id
+    }
+}
+
+impl Eq for Mapping {}
+
+impl Borrow<ListenerId> for Mapping {
+    fn borrow(&self) -> &ListenerId {
+        &self.listener_id
+    }
+}
+
+/// Current state of a [`Mapping`].
+enum MappingState {
+    /// Port mapping is inactive, will be requested or re-requested on the next iteration.
+    Inactive,
+    /// Port mapping/removal has been requested on the gateway.
+    Pending,
+    /// Port mapping is active with the inner timeout.
+    Active(Delay),
+    /// Port mapping is permanent on the Gateway.
+    Permanent,
+}
+
+/// Current state of the UPnP [`Provider::Gateway`].
+enum GatewayState<P: Provider> {
+    Searching(BoxFuture<'static, Result<(P::Gateway, Ipv4Addr), Box<dyn std::error::Error>>>),
+    Available((Arc<P::Gateway>, Ipv4Addr)),
+    GatewayNotFound,
+}
+
+/// A `NetworkBehaviour` for UPnP port mapping. Automatically tries to map the external port
+/// to an internal address on the gateway on a `FromSwarm::NewListenAddr`.
+pub struct Behaviour<P>
+where
+    P: Provider,
+{
+    /// Gateway config.
+    config: Config,
+    /// UPnP interface state.
+    state: GatewayState<P>,
+
+    /// List of port mappings.
+    mappings: HashMap<Mapping, MappingState>,
+
+    /// Pending gateway events.
+    pending_events: FuturesUnordered<BoxFuture<'static, Event>>,
```

Comment:
This will poll everything on the main behaviour task. We should avoid doing IO on that one and instead spawn a worker task and talk to it over a channel!

Reply:
ok thanks Thomas. In this case two channels right? One to send the tasks and one to poll for the completed futures. Do you have an example for that on any other protocol? (Couldn't find one). We should also add a `spawn` function to the `Provider` trait right?

Reply:
Yeah, you'll need two channels, true.

I don't think we have that anywhere else and yes you'll need a `spawn` function.

Reply:
Technically, you could spawn each search and only pass a oneshot for the result in.

That would probably be simpler and I don't think we need to be worried about the number of tasks or unbounded growth because we only trigger this on new listen addresses.

Reply:
it's also triggered on port mapping renewal. I updated `Behaviour` with two channels, one `Unbounded` to send the futures to the task, and one bounded to send the `Event`s back to the `Behaviour`. I don't see a problem here with the `UnboundedChannel` as `FuturesUnordered` is also "unbounded" but CC @mxinden. I also had to update `Gateway` functions signatures to `String` instead of `Box<dyn std::error::Error>` as it's not send ptal Thomas.